Stale Hickory Goblin

Medium

# The return value of the `convertAndDeposit()` function is not checked

## Summary

The `convertAndDeposit()` call `vault.delegateCall()`, which in turn calls `delegatecall`, but the return value (`success`) is not checked.

## Vulnerability Detail

In `SimpleDVTStakingStrategy.sol` we have `convertAndDeposit()` function:
```solidity
    function convertAndDeposit(
        uint256 amount,
        uint256 blockNumber,
        bytes32 blockHash,
        bytes32 depositRoot,
        uint256 nonce,
        bytes calldata depositCalldata,
        IDepositSecurityModule.Signature[] calldata sortedGuardianSignatures
    ) external returns (bool success) {
        (success, ) = vault.delegateCall( //@audit - retun value not checked
            address(stakingModule),
            abi.encodeWithSelector(
                IStakingModule.convertAndDeposit.selector,
                amount,
                blockNumber,
                blockHash,
                depositRoot,
                nonce,
                depositCalldata,
                sortedGuardianSignatures
            )
        );
        emit ConvertAndDeposit(success, msg.sender);
    }
```

This function interact with a staking module to convert a specified amount of tokens and then deposit them. 

The function uses a delegate call to the staking module, ensuring that the operation is executed within the context of the vault.

```solidity
(success, ) = vault.delegateCall(
```
The result of the delegate call is captured in the `success` variable. This indicates whether the operation was successful.

This `delegatecall` is calling from `delegateCall()` in `Vault.sol`:
```solidity
(success, response) = to.delegatecall(data);
```

However, the `success` variable is not checked for being successful in both functions. 

According to the Solidity [documentation](https://docs.soliditylang.org/en/develop/control-structures.html#error-handling-assert-require-revert-and-exceptions):

> The low-level functions `call`, `delegatecall` and `staticcall` return `true` as their first return value if the account called is non-existent, as part of the design of the EVM. Account existence must be checked prior to calling if needed.
## Impact

Тhere is a possibility that the call wasn't successful, while the transaction continues without reverting.

## Code Snippet

https://github.com/sherlock-audit/2024-06-mellow/blob/main/mellow-lrt/src/strategies/SimpleDVTStakingStrategy.sol#L45
https://github.com/sherlock-audit/2024-06-mellow/blob/main/mellow-lrt/src/Vault.sol#L280

## Tool used

Manual Review

## Recommendation

Check if the transaction was successful in `convertAndDeposit()`.